src/main/java/it/gov/pagopa/reward/notification/service/iban/outcome/recovery/BaseDiscardedRewardNotificationServiceImpl.java

```diff
@@ -0,0 +1,69 @@
+
+import java.time.LocalDate;
+
+public abstract class BaseDiscardedRewardNotificationServiceImpl implements DiscardedRewardNotificationService{
+
+    private final RewardNotificationRuleService notificationRuleService;
+
+
+    // handlers to calculate notificationDate
+    private final RewardNotificationTemporalHandlerServiceImpl temporalHandler;
+    private final RewardNotificationBudgetExhaustedHandlerServiceImpl budgetExhaustedHandler;
+    private final RewardNotificationThresholdHandlerServiceImpl thresholdHandler;
+
+    protected BaseDiscardedRewardNotificationServiceImpl(RewardNotificationRuleService notificationRuleService,
+                                                         RewardNotificationTemporalHandlerServiceImpl temporalHandler,
+                                                         RewardNotificationBudgetExhaustedHandlerServiceImpl budgetExhaustedHandler,
+                                                         RewardNotificationThresholdHandlerServiceImpl thresholdHandler) {
+        this.notificationRuleService = notificationRuleService;
+        this.temporalHandler = temporalHandler;
+        this.budgetExhaustedHandler = budgetExhaustedHandler;
+        this.thresholdHandler = thresholdHandler;
+    }
+
+    @Override
+    public Mono<RewardsNotification> setRemedialNotificationDate(String initiativeId, RewardsNotification notification) {
+
+        return notificationRuleService.findById(initiativeId)
+                .map(rule -> {
+                    LocalDate today = LocalDate.now();
+
+                    if (rule.getEndDate() != null && rule.getEndDate().isBefore(today)) {
+                        return budgetExhaustedHandler.calculateNotificationDate();
+                    } else if(rule.getTimeParameter() != null){
+                        return temporalHandler.calculateNotificationDate(today, rule);
+                    } else if(rule.getAccumulatedAmount() != null){
+                        if(AccumulatedAmountDTO.AccumulatedTypeEnum.BUDGET_EXHAUSTED.equals(rule.getAccumulatedAmount().getAccumulatedType())){
+                            return budgetExhaustedHandler.calculateNotificationDate();
+                        } else if(rule.getAccumulatedAmount().getRefundThreshold() != null)  {
+                            return thresholdHandler.calculateNotificationDate();
+                        } else {
+                            throw new IllegalStateException("Not valid threshold rule %s".formatted(rule.getInitiativeId()));
+                        }
+                    } else {
+                        throw new IllegalStateException("Not valid rule %s".formatted(rule.getInitiativeId()));
+                    }
+                })
+                .map(date -> {
+                    notification.setNotificationDate(date);
+                    return notification;
+                });
+    }
+
+    public static void resetNotificationStatus(RewardsNotification rewardsNotification) {
```

Comment:
Maybe actually we could not put this method as static public, and put it inside its Service